src/driver-marsgaming/marsgaming-probe.c

```diff
@@ -0,0 +1,159 @@
+#include "marsgaming-probe.h"
+
+#include "marsgaming-definitions.h"
+#include "marsgaming-query.h"
+
+static void
+marsgaming_probe_button_type(struct ratbag_button *button)
+{
+	button->type = marsgaming_button_type(button->index);
+}
+
+static void
+marsgaming_probe_button_action(struct ratbag_button *button,
+			       const struct marsgaming_button_info *button_info)
+{
+	struct ratbag_button_action action = marsgaming_parse_button_to_action(button, button_info);
+	ratbag_button_set_action(button, &action);
+}
+
+static void
+marsgaming_probe_profile_leds(struct ratbag_profile *profile)
+{
+	struct marsgaming_profile_drv_data *profile_data = marsgaming_profile_get_drv_data(profile);
+	profile_data->led_report = marsgaming_query_profile_led(profile);
+	struct ratbag_led *led;
+	ratbag_profile_for_each_led(profile, led) {
+		led->type = RATBAG_LED_TYPE_LOGO;
+		ratbag_led_set_mode_capability(led, RATBAG_LED_OFF);
+		ratbag_led_set_mode_capability(led, RATBAG_LED_ON);
+		ratbag_led_set_mode_capability(led, RATBAG_LED_BREATHING);
+		led->colordepth = RATBAG_LED_COLORDEPTH_RGB_888;
+		struct marsgaming_report_led report = profile_data->led_report;
+		led->color = marsgaming_led_color_to_ratbag(report.led.color);
+		led->brightness = report.led.brightness * (255 / 3);
+		if (report.led.brightness == 0) {
+			led->mode = RATBAG_LED_OFF;
+		} else if (report.led.breathing_speed == 0 || report.led.breathing_speed >= 10) {
+			led->mode = RATBAG_LED_ON;
+		} else { // Breathing mode
+			led->mode = RATBAG_LED_BREATHING;
+			led->ms = report.led.breathing_speed * 2000;
+		}
+	}
+}
+
+static void
+marsgaming_probe_button(struct ratbag_button *button, struct marsgaming_button_info *button_info)
+{
+	ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_BUTTON);
+	ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_SPECIAL);
+	ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_KEY); // Deprecated
```

Comment:
Why `Deprecated`?

Reply:
I was told [here](https://github.com/libratbag/libratbag/issues/1199#issuecomment-923089817) that `key` functionality was deprecated. I will remove this line, since it isn't even implemented.